File: src/tech/octopusdragon/intervaltimer/Alert.java
package tech.octopusdragon.intervaltimer;

import javax.sound.midi.MetaEventListener;
import javax.sound.midi.MetaMessage;
import javax.sound.midi.MidiSystem;
import javax.sound.midi.Sequencer;

/**
 * The Alert class is used to encapsulate the alerts that can be used.
 * @author Alex Gill
 *
 */

public enum Alert {
	
	TELEPHONE("telephone.mid"),
	CHIMES_UP("chimes_up.mid"),
	CHIMES_DOWN("chimes_down.mid"),
	WHISTLE("whistle.mid"),
	WOODBLOCK("woodblock.mid"),
	XYLOPHONE("xylophone.mid"),
	OCARINA("ocarina.mid"),
	GRANDFATHER_1("grandfather_clock1.mid"),
	GRANDFATHER_2("grandfather_clock2.mid"),
	DING("ding.mid"),
	DING_DONG("ding_dong.mid"),
	DONG_DING("dong_ding.mid"),
	CYMBAL("cymbal.mid"),
	APITO("apito.mid"),
	CHORDS("chords.mid"),
	OBATALA("obatala.mid"),
	BRAVE_WIND("brave_wind.mid"),
	OMOMUKI("omomuki.mid"),
	SILENT();
	
	// --- Constants ---
	private final static String ALERT_DIR = "resources/alerts/";
	
	// --- Variables ---
	private Sequencer player;		// The Sequencer for the Alert
	private String filename;		// The filename of the alert audio file
	
	/**
	 * Instantiates a new silent Alert
	 */
	private Alert() {
	}
	
	
	/**
	 * Instantiates a new Alert
	 * @param The filename of the alert
	 */
	private Alert(String filename) {
		this.filename = ALERT_DIR + filename;
	}
	
	
	/**
	 * Plays the alert sound
	 */
	public void play() {
		if (filename == null) return;
		
		// Create a new Sequencer and set its Sequence to the Alert.
		try {
			if (player != null)
				player.close();
			player = MidiSystem.getSequencer();
			player.addMetaEventListener(new EndOfSequenceListener());
			player.open();
			player.setSequence(getClass().getClassLoader().getResourceAsStream(
					filename));
			player.start();
			
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	
	/**
	 * Listener for the end of the Sequence
	 * @author Alex Gill
	 *
	 */
	class EndOfSequenceListener implements MetaEventListener {
		@Override
		public void meta(MetaMessage message) {
			
			// This code will execute once the end of the Sequence is reached
			if (message.getType() == 47) {
				player.close();
			}
		}
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder(this.name().toLowerCase());
		// Replace underscores with spaces
		for (int i = 0; i < sb.length(); i++) {
			if (sb.charAt(i) == '_') {
				sb.replace(i, i + 1, " ");
			}
		}
		// Capitalize
		sb.replace(0, 1, sb.substring(0, 1).toUpperCase());
		for (int i = 0; i < sb.length(); i++) {
			if (i >= 1 && sb.charAt(i - 1) == ' ') {
				sb.replace(i, i + 1, sb.substring(i, i + 1).toUpperCase());
			}
		}
		return sb.toString();
	}
}

File: constants/Alert.ts
import { Audio } from "expo-av";

const dir = "../assets/alerts/";

const telephone_file = require(dir + "telephone.mp3");
const chimes_up_file = require(dir + "chimes_up.mp3");
const chimes_down_file = require(dir + "chimes_down.mp3");
const whistle_file = require(dir + "whistle.mp3");
const woodblock_file = require(dir + "woodblock.mp3");
const xylophone_file = require(dir + "xylophone.mp3");
const ocarina_file = require(dir + "ocarina.mp3");
const grandfather_1_file = require(dir + "grandfather_clock1.mp3");
const grandfather_2_file = require(dir + "grandfather_clock2.mp3");
const ding_file = require(dir + "ding.mp3");
const ding_dong_file = require(dir + "ding_dong.mp3");
const dong_ding_file = require(dir + "dong_ding.mp3");
const cymbal_file = require(dir + "cymbal.mp3");
const apito_file = require(dir + "apito.mp3");
const chords_file = require(dir + "chords.mp3");
const obatala_file = require(dir + "obatala.mp3");
const brave_wind_file = require(dir + "brave_wind.mp3");
const omomuki_file = require(dir + "omomuki.mp3");

export default class Alert {

	static values: Alert[] = [];
	static map: {[id: string]: Alert} = {};

	static TELEPHONE = new Alert("Telephone", telephone_file);
	static CHIMES_UP = new Alert("Chime Up", chimes_up_file);
	static CHIMES_DOWN = new Alert("Chime Down", chimes_down_file);
	static WHISTLE = new Alert("Whistle", whistle_file);
	static WOODBLOCK = new Alert("Woodblock", woodblock_file);
	static XYLOPHONE = new Alert("Xylophone", xylophone_file);
	static OCARINA = new Alert("Ocarina", ocarina_file);
	static GRANDFATHER_1 = new Alert("Grandf. 1", grandfather_1_file);
	static GRANDFATHER_2 = new Alert("Grandf. 2", grandfather_2_file);
	static DING = new Alert("Ding", ding_file);
	static DING_DONG = new Alert("Ding Dong", ding_dong_file);
	static DONG_DING = new Alert("Dong Ding", dong_ding_file);
	static CYMBAL = new Alert("Cymbal", cymbal_file);
	static APITO = new Alert("Apito", apito_file);
	static CHORDS = new Alert("Chords", chords_file);
	static OBATALA = new Alert("Obatala", obatala_file);
	static BRAVE_WIND = new Alert("Brave Wind", brave_wind_file);
	static OMOMUKI = new Alert("Omomuki", omomuki_file);
	static SILENT = new Alert("Silent");

	static sound: Audio.Sound = new Audio.Sound();

  name: string;
  file: any;

  private constructor(name: string, file?: any) {
    this.name = name;
    this.file = file ? file : undefined;
		Alert.values.push(this);
		Alert.map[name] = this;
  }
	
  async play() {
    if (!this.file) return;
    await Alert.sound.unloadAsync();
    await Alert.sound.loadAsync(this.file);
    await Alert.sound.playAsync();
  }
}